### PR TITLE
Introduce the concept of "tasks" to prepare for different evaluation tasks like "write tests" and "repair code"

### PR DESCRIPTION
**Pull request by ahumenberger**
Part of #165

### CODE REVIEW

File: task/task.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+func registerIdentifier(name string) (identifier Identifier) {
+	identifier = Identifier(name)
+	AllIdentifiers = append(AllIdentifiers, identifier)
+
+	return identifier
+}
+
+var (
+	// IdentifierWriteTests holds the identifier for the "write test" task.
+	IdentifierWriteTests = registerIdentifier("write-tests")
```
Comment (ahumenberger) [inline task/task.go:31]: Not sure about the naming. I think before it was called "generate tests", I just noticed that in the end.

**Comment by ahumenberger** [comment]
I think it worth doing https://github.com/symflower/eval-dev-quality/issues/169 first.

File: task/task.go (hunk 1 above)
Comment (zimmski) [reply task/task.go:36]: write tests is fine with me. Since the POV is that we are benchmarking a developer.

Just to note this: in the LLM world we usually talk about CodeGen

File: task/task.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	"github.com/symflower/eval-dev-quality/language"
+)
+
+var (
+	// ErrTaskUnsupported indicates that a task is unsupported.
+	ErrTaskUnsupported = errors.New("task unsupported")
+)
+
+// Identifier holds the identifier of a task.
+type Identifier string
```
Comment (zimmski) [inline task/task.go:16]: I like this to be reworded. In evaluate/metrics/assessment.go we have "Key". @bauersimon @ahumenberger any other suggestions? "Identifier" doesn't sound right to me.

File: evaluate/evaluate_test.go [hunk 3]
```diff
@@ -27,6 +27,7 @@ import (
 	modeltesting "github.com/symflower/eval-dev-quality/model/testing"
 	"github.com/symflower/eval-dev-quality/provider"
 	providertesting "github.com/symflower/eval-dev-quality/provider/testing"
+	"github.com/symflower/eval-dev-quality/task"
```
Comment (zimmski) [inline evaluate/evaluate_test.go:30]: Please update the README as well with these changes. We now have a generic task interface.

File: model/symflower/symflower.go [hunk 4]
```diff
@@ -33,27 +32,47 @@ func (m *Model) ID() (id string) {
 	return "symflower" + provider.ProviderModelSeparator + "symbolic-execution"
 }
 
-// GenerateTestsForFile generates test files for the given implementation file in a repository.
-func (m *Model) GenerateTestsForFile(logger *log.Logger, language language.Language, repositoryPath string, filePath string) (assessment metrics.Assessments, err error) {
```
Comment (zimmski) [inline model/symflower/symflower.go:37]: Please keep that exported function. Fine with converting it to have context parameter, but unexporting it makes using it impossible in other code. And i would like to rework the interface when i am done with the report i need to write. That would mean reverting that as well.

File: model/model.go [hunk 5]
```diff
@@ -2,17 +2,18 @@ package model
...
+	"github.com/symflower/eval-dev-quality/task"
 )
 
 // Model defines a model that can be queried for generations.
 type Model interface {
 	// ID returns the unique ID of this model.
 	ID() (id string)
 
-	// GenerateTestsForFile generates test files for the given implementation file in a repository.
-	GenerateTestsForFile(logger *log.Logger, language language.Language, repositoryPath string, filePath string) (assessments metrics.Assessments, err error)
```
Comment (zimmski) [inline model/model.go:15]: Same as with https://github.com/symflower/eval-dev-quality/pull/166/commits/4e6d0619eb159f7212ab4a713f45d0b404b92271#r1632828265

File: evaluate/report/csv_test.go [hunk 6]
```diff
@@ -32,57 +33,63 @@ func TestGenerateCSVForAssessmentPerModelPerLanguagePerRepository(t *testing.T)
...
-					"some-repository": metrics.Assessments{
-						metrics.AssessmentKeyGenerateTestsForFileCharacterCount: 50,
-						metrics.AssessmentKeyResponseCharacterCount:             100,
-						metrics.AssessmentKeyCoverage:                           1,
-						metrics.AssessmentKeyFilesExecuted:                      2,
-						metrics.AssessmentKeyResponseNoError:                    3,
-						metrics.AssessmentKeyResponseNoExcess:                   4,
-						metrics.AssessmentKeyResponseWithCode:                   5,
-						metrics.AssessmentKeyProcessingTime:                     200,
+					"some-repository": {
```
Comment (zimmski) [inline evaluate/report/csv_test.go:57]: Please use the type for composite literals.

File: task/task.go (hunk 2 above)
Comment (bauersimon) [reply task/task.go:16]: Each Model, Provider, Tool on the other hand does have an `ID`. But that is a slightly different structure.

Comment (bauersimon) [reply task/task.go:16]: Or since it is the only thing that currently identifies what a task is, why not call it `Task`?

Comment (bauersimon) [reply task/task.go:16]: And then `AllIdentifiers` -> `AllTasks`.

Comment (zimmski) [reply task/task.go:16]: I see. That brings me back to my original request i had with Andi: making the tasks use methods of models/languages instead of letting the models reimplement tasks. Then every task has the same pattern as model/language/provider/...

File: model/symflower/symflower.go (hunk 4 above)
Comment (ahumenberger) [reply model/symflower/symflower.go:37]: Can do that, but I don't think that's how we usually work, unless there is a concrete plan how this will look like in the end. Feels like YAGNI.

> That would mean reverting that as well.

Reverting needs to be done anyhow since the change is already done in this PR.

> And i would like to rework the interface 

Please do a planning for this. I also put some thought into this structure.

File: task/task.go (hunk 2 above)
Comment (ahumenberger) [reply task/task.go:16]: The assessment type is basically a database, so "Key" fits perfectly there, but it does not for tasks. A task identifier identifies a task. E.g. a repository can  list tasks which are supposed to be evaluated. Sounds right to me.

Comment (ahumenberger) [reply task/task.go:16]: > making the tasks use methods of models/languages instead of letting the models reimplement tasks.

This is meant to be done as a follow-up. This PR only introduces the primitive concept running different tasks. For this, a "task identifier" is sufficient. Later a `Task` type can be introduced. A "task identifier" is still necessary, because e.g. a repository can specify a list of tasks that should be run.

File: model/llm/llm.go [hunk 7]
```diff
@@ -32,7 +32,7 @@ type Model struct {
 }
 
 // NewModel returns an LLM model corresponding to the given identifier which is queried via the given provider.
-func NewModel(provider provider.Query, modelIdentifier string) model.Model {
+func NewModel(provider provider.Query, modelIdentifier string) *Model {
```
Comment (bauersimon) [inline model/llm/llm.go:35]: Not a huge thing but why is this necessary?

File: evaluate/repository.go [hunk 8]
```diff
@@ -19,34 +19,42 @@ import (
...
 )
 
-// Repository evaluate a repository with the given model and language.
-func Repository(logger *log.Logger, resultPath string, model evalmodel.Model, language language.Language, testDataPath string, repositoryName string) (repositoryAssessment metrics.Assessments, problems []error, err error) {
-	log, logClose, err := log.WithFile(logger, filepath.Join(resultPath, evalmodel.CleanModelNameForFileSystem(model.ID()), language.ID(), repositoryName+".log"))
+// Repository holds data about a repository.
+type Repository struct {
+	// Name holds the name of the repository.
+	Name string
+	// DataPath holds the path to the repository.
```
Comment (bauersimon) [inline evaluate/repository.go:54]: Absolute or relative? Just asking because we usually also have this `TestdataPath` that makes all the repository paths relative.

File: evaluate/repository.go [hunk 9]
```diff
@@ -19,34 +19,42 @@ import (
...
 	if err != nil {
 		return nil, problems, pkgerrors.WithStack(err)
 	}
 
 	repositoryAssessment = metrics.NewAssessments()
 	for _, filePath := range filePaths {
-		if err := ResetTemporaryRepository(logger, testDataPath); err != nil {
+		if err := r.Reset(logger); err != nil {
 			logger.Panicf("ERROR: unable to reset temporary repository path: %s", err)
 		}
```
Comment (bauersimon) [inline evaluate/repository.go:100]: Usability nitpick: since we anyways panic here (and I think we always panic if the reset mechanism fails) we could also just panic within `Reset`.

File: evaluate/repository.go [hunk 10]
```diff
@@ -19,16 +20,63 @@ import (
...
+func (rc *repositoryConfiguration) validate() (err error) {
+	if len(rc.Tasks) == 0 {
+		return pkgerrors.Errorf("empty list of tasks in configuration")
+	}
+
+	for _, taskIdentifier := range rc.Tasks {
+		if !task.LookupIdentifier[taskIdentifier] {
+			return pkgerrors.Errorf("task identifier %q unknown", taskIdentifier)
+		}
+	}
```
Comment (bauersimon) [inline evaluate/repository.go:38]: Plz wrap errors with stacks.

File: evaluate/evaluate_test.go [hunk 11]
```diff
@@ -535,11 +559,12 @@ func TestEvaluate(t *testing.T) {
...
-						languageGolang: map[string]metrics.Assessments{
-							repositoryPlainPath: map[metrics.AssessmentKey]uint64{},
-							repositoryNextPath:  map[metrics.AssessmentKey]uint64{},
+				ExpectedAssessments: map[evalmodel.Model]map[language.Language]map[string]map[task.Identifier]metrics.Assessments{
+					mockedModel: map[language.Language]map[string]map[task.Identifier]metrics.Assessments{
+						languageGolang: map[string]map[task.Identifier]metrics.Assessments{
+							repositoryPlainPath: map[task.Identifier]metrics.Assessments{
+								task.IdentifierWriteTests: map[metrics.AssessmentKey]uint64{},
+							},
 						},
```
Comment (bauersimon) [inline evaluate/evaluate_test.go:568]: `repositoryNextPath` is gone. Guess does not matter if its empty or gone, but I think we should keep it?

File: evaluate/evaluate_test.go [hunk 12]
```diff
@@ -121,10 +121,12 @@ func TestEvaluate(t *testing.T) {
...
-					for _, a := range rs {
-						if v, ok := a[metrics.AssessmentKeyProcessingTime]; ok {
-							if assert.Greater(t, v, uint64(0)) {
-								delete(a, metrics.AssessmentKeyProcessingTime)
+					for _, ts := range rs {
+						for _, a := range ts {
+							if v, ok := a[metrics.AssessmentKeyProcessingTime]; ok {
+								if assert.Greater(t, v, uint64(0)) {
+									delete(a, metrics.AssessmentKeyProcessingTime)
+								}
```
Comment (bauersimon) [inline evaluate/evaluate_test.go:129]: Out of context.

We already had a logic that ignores all nondeterministic parts of assessments. So with https://github.com/symflower/eval-dev-quality/commit/b207e284f4a76054dd3e2bf4259f7d2e744bffe7 we now do that twice in two separate locations. Could you please either resolve this or keep it as a follow-up?

File: evaluate/repository.go [hunk 13]
```diff
@@ -19,16 +20,63 @@ import (
...
+
+	return r.repositoryConfiguration.validate()
+}
+
 // Evaluate evaluates a repository with the given model and language.
-func (r *Repository) Evaluate(logger *log.Logger, resultPath string, model evalmodel.Model, language language.Language) (repositoryAssessment metrics.Assessments, problems []error, err error) {
+func (r *Repository) Evaluate(logger *log.Logger, resultPath string, model evalmodel.Model, language language.Language, taskIdentifier task.Identifier) (repositoryAssessment metrics.Assessments, problems []error, err error) {
 	log, logClose, err := log.WithFile(logger, filepath.Join(resultPath, evalmodel.CleanModelNameForFileSystem(model.ID()), language.ID(), r.Name+".log"))
 	if err != nil {
 		return nil, nil, err
```
Comment (bauersimon) [inline evaluate/repository.go:82]: Please  consider changing the logging to also incorporate the task, and add the task also to the log call below (which I cannot select for review - thanks GitHub!).

**Review by bauersimon — CHANGES_REQUESTED** [review]

File: model/llm/llm.go (hunk 7 above)
Comment (ahumenberger) [reply model/llm/llm.go:35]: Because in the tests we create a model with `NewModel` and then call `GenerateTestsFor....` and this is not a method of `model.Model` anymore.

File: evaluate/evaluate_test.go (hunk 11 above)
Comment (ahumenberger) [reply evaluate/evaluate_test.go:568]: Why do you think we should keep it?

The test is the following: "Never solving basic checks leads to exclusion". So this is IMO fine if there is nothing for `repositoryNextPath`.

Also the reason why it was there before is because the nested maps were preinitialized for all models, languages and repositories. We would need to do the same for tasks. But not every model supports every task, and not every repository allows all tasks. So this gets complicated to preinitialize. It's way easier to initialize on demand. So unless there is a good reason why we need the old version, I'd keep  this one.

File: task/task.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+// AllIdentifiers holds all available task identifiers.
+var AllIdentifiers []Identifier
+
+// registerIdentifier registers the given identifier and makes it available.
+func registerIdentifier(name string) (identifier Identifier) {
+	identifier = Identifier(name)
+	AllIdentifiers = append(AllIdentifiers, identifier)
+
+	return identifier
+}
```
[the remark this answers is not in the document]
Comment (ahumenberger) [reply task/task.go]: In a later commit I'm introducing a map as well. Added a check now to make sure that identifiers are not added twice.

File: evaluate/repository.go (hunk 10 above)
Comment (bauersimon) [reply evaluate/repository.go:38]: ahhhh I confused that one.

File: evaluate/repository.go (hunk 9 above)
Comment (ahumenberger) [reply evaluate/repository.go:100]: Nah, not a friend of that. It should be the caller's decision what to do with the error.

File: evaluate/report/csv_test.go (hunk 6 above)
Comment (ahumenberger) [reply evaluate/report/csv_test.go:57]: The tests were already like this. And since I'm planning on changing this with #169 I'm not really eager to change this now.

File: evaluate/repository.go (hunk 8 above)
Comment (ahumenberger) [reply evaluate/repository.go:54]: absolute. changed the docstring for the field

File: evaluate/evaluate_test.go (hunk 12 above)
Comment (ahumenberger) [reply evaluate/evaluate_test.go:129]: Not sure what you mean. If you are talking about the "normalize assessments, then this is the very same location.

File: model/model.go (hunk 5 above)
Comment (ahumenberger) [reply model/model.go:15]: See https://github.com/symflower/eval-dev-quality/pull/166#discussion_r1634653364

File: task/task.go (hunk 2 above)
Comment (bauersimon) [reply task/task.go:16]: Then let's do `Key` for now? And when we touch it again, we can see what makes most sense then.

Comment (ahumenberger) [reply task/task.go:16]: Since I think "Identifier" is totally fine, and there are no better suggestions, I'm going to leave it.

File: evaluate/evaluate_test.go (hunk 12 above)
Comment (bauersimon) [reply evaluate/evaluate_test.go:129]: It is... sorry for not clarifying. There already is `metricstesting.AssertAssessmentsEqual`. And now this as well.

File: evaluate/repository.go (hunk 13 above)
Comment (bauersimon) [reply evaluate/repository.go]: Should we maybe change the log file location as well? I.e. different files for different tasks, otherwise they will clump up in one log file.

File: evaluate/evaluate_test.go (hunk 12 above)
Comment (bauersimon) [reply evaluate/evaluate_test.go:129]: its not a big deal... just good to keep an eye out in the future

File: evaluate/repository.go (hunk 13 above)
Comment (ahumenberger) [reply evaluate/repository.go]: true true true

File: evaluate/evaluate_test.go (hunk 12 above)
Comment (ahumenberger) [reply evaluate/evaluate_test.go:129]: Briefly looked at this. But I don't want to touch this right now. Might be that some tests are failing again, and this would delay merging the PR.

Comment (ahumenberger) [reply evaluate/evaluate_test.go:129]: Added to #169

**Review by bauersimon — APPROVED** [review]
